File: leetcode/python/0200. Number of Islands.py
# Time complexity: O(RC) where R is the num of rows, C is the number of cols
# Space complexity: O(RC) where R is the num of rows, C is the number of cols

class Solution(object):
    def numIslands(self, grid):
        row = len(grid)
        col = len(grid[0])
        islands = 0

        visited = [[0 for _ in range(col)] for _ in range(row)]
        
        for r in range(row):
            for c in range(col):
                if visited[r][c] == 0 and grid[r][c] == "1":
                    self.dfs(r,c,visited, grid)
                    islands += 1
        return islands

    def dfs(self, row, col, visited, grid):
        r = len(visited)
        c = len(visited[0])

        if row >=0 and col>=0 and row<r and col<c and visited[row][col] == 0 and grid[row][col] == "1":
            visited[row][col] = 1
            self.dfs(row-1, col, visited, grid)
            self.dfs(row+1, col, visited, grid)
            self.dfs(row, col-1, visited, grid)
            self.dfs(row, col+1, visited, grid)

File: leetcode/java/0200. Number of Islands.java
package leetcode.java;

// Time complexity: O(RC) where R is the num of rows, C is the number of cols
// Space complexity: O(RC) where R is the num of rows, C is the number of cols

class Solution {
    public int numIslands(char[][] grid) {
        int numIslands = 0;

        for (int row=0; row<grid.length; row++) {
            for (int col=0; col<grid[0].length; col++) {
                if (grid[row][col] == '1') {
                    numIslands += 1;
                    this.dfs(grid, row, col);
                }
            }
        }
        return numIslands;
    }

    public void dfs(char[][] grid, int row, int col) {
        
        if (row >= 0 && row <grid.length && col >= 0 && col < grid[0].length && grid[row][col] == '1') {
            grid[row][col] = 'a';
            this.dfs(grid, row+1, col);
            this.dfs(grid, row-1, col);
            this.dfs(grid, row, col-1);
            this.dfs(grid, row, col+1);
        }
    }
}